plugins/org.eclipse.fordiac.ide.library/src/org/eclipse/fordiac/ide/library/LibraryManager.java

```diff
@@ -834,125 +712,300 @@ public IStatus runInWorkspace(final IProgressMonitor monitor) throws CoreExcepti
 
-		final IFolder typelib = project.getFolder(TYPE_LIB_FOLDER_NAME);
-		try {
-			for (final var member : typelib.members()) {
-				if (member instanceof final IContainer container) {
-					final Manifest libManifest = ManifestHelper.getContainerManifest(container);
-					if (libManifest != null && resolved.contains(libManifest.getProduct().getSymbolicName())
-							&& libManifest.getDependencies() != null) {
-						libManifest.getDependencies().getRequired().forEach(req -> handleRequired(requirements, causes,
-								projectManifest, markerList, req, libManifest));
+		while (!queue.isEmpty()) {
+			final String symbolicName = queue.poll();
+
+			final var dnode = deps.get(symbolicName);
+
+			if (!dnode.isChanged()) {
+				continue;
+			}
+
+			if (!dnode.isValid()) {
+				final var rnode = res.get(symbolicName);
+				if (rnode != null) {
+					rnode.getDependencies().keySet().forEach(symb -> {
+						final var dn = deps.get(symb);
+						if (dn != null) {
+							dn.removeCause(symbolicName);
+							if (dn.isChanged()) {
+								queue.add(symb);
+							}
+						}
+					});
+				}
+				continue;
+			}
+
+			// resolve dependency
+			final var rnode = resolveDependency(symbolicName, dnode.getRange(), preferred.get(symbolicName));
+
+			if (res.containsKey(symbolicName)) {
+				final var oldRNode = res.get(symbolicName);
+				oldRNode.getDependencies().keySet().forEach(old -> {
+					if (!rnode.getDependencies().containsKey(old)) {
+						deps.get(old).removeCause(symbolicName);
+					}
+				});
+			}
+
+			res.put(symbolicName, rnode);
+
+			// updated dependencies
+			rnode.getDependencies().forEach((symb, val) -> {
+				final var dn = deps.computeIfAbsent(symb, s -> new DependencyNode(symb));
+				dn.putCause(symbolicName, val);
+				if (dn.isChanged()) {
+					queue.add(symb);
+				}
+			});
+		}
+
+		final List<ErrorMarkerBuilder> markerList = new LinkedList<>();
+
+		// import valid nodes
+		for (final var dnode : deps.values()) {
+			if (dnode.isValid()) {
+				final var rnode = res.get(dnode.getSymbolicName());
+
+				if (rnode.isValid()) {
+					// no need to import the same version again
+					if (!linked.contains(rnode.getSymbolicName())
+							|| !preferred.get(rnode.getSymbolicName()).equals(rnode.getVersion())) {
+						importLibrary(project, typeLibrary, rnode.getUri(), false, false);
+
 					}
+					linked.remove(rnode.getSymbolicName());
+				} else {
+					markerList.add(createDependencyMarker(projectManifest, rnode, dnode));
+
 				}
+			} else if (dnode.isRangeEmpty()) {
+				markerList.add(createDependencyMarker(projectManifest, dnode));
+			}
+		}
+
+		linked.forEach(l -> {
+			try {
+				project.getFolder(TYPE_LIB_FOLDER_NAME).getFolder(l).delete(true, null);
+			} catch (final CoreException e) {
+				// empty
 			}
+		}); // remove still linked libraries
+
+		FordiacMarkerHelper.updateMarkers(project.getFile(MANIFEST), FordiacErrorMarker.LIBRARY_MARKER, markerList,
+				true);
```

Comment:
I would suggest to create the markers directly, since we are already running in a workspace job:
```suggestion
		final IFile manifestFile = project.getFile(MANIFEST);
		if (manifestFile.isAccessible()) {
			try {
				manifestFile.deleteMarkers(FordiacErrorMarker.LIBRARY_MARKER, true, IResource.DEPTH_INFINITE);
				for (final var builder : markerList) {
					builder.createMarker(manifestFile);
				}
			} catch (final CoreException e) {
				FordiacLogHelper.logError("Could not update error markers on resource: " + manifestFile.getName(), //$NON-NLS-1$
						e);
			}
		}
```